### PR TITLE
Test coverage for REST API pkp/pkp-lib#2875

### PR DESCRIPTION
**Pull request by kaschioudi**

### CODE REVIEW

File: tests/api/SubmissionsTest.php [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+ */
+
+import('lib.pkp.tests.PKPApiTestCase');
+
+class SubmissionsTest extends PKPApiTestCase {
+	/**
+	 * @covers /submissions
+	 * @expectedException GuzzleHttp\Exception\ClientException
+	 */
+	public function testGetSubmissionsWithoutToken() {
```
Comment (asmecher) [inline tests/api/SubmissionsTest.php]: It looks like this test (and some of the tests below) could move into the `pkp-lib` submodule, so they'd apply to OMP as well.

File: tests/api/SubmissionsTest.php [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+
+	/**
+	 * @covers /submissions/{submissionId}
+	 */
+	public function testGetSubmissionByIdWithValidId() {
+		$response = $this->_sendRequest('GET', '/submissions/25');
+		$this->assertSame(200, $response->getStatusCode());
+
+		$data = $this->_getResponseData($response);
+		$this->assertArrayHasKey('id', $data);
```
Comment (asmecher) [inline tests/api/SubmissionsTest.php]: These kinds of assertions start to duplicate the Swagger document. I wonder whether something like https://github.com/nabbar/SwaggerValidator-PHP wouldn't allow us to write tests here but still delegate the validation (e.g. required fields) to the Swagger schema.

File: tests/api/SubmissionsTest.php [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+	public function testGetSubmissionParticipantsAssignedToStageWithoutToken() {
+		$response = $this->_sendRequest('GET', '/submissions/25/participants/1', array(), false);
+		$this->assertSame(404, $response->getStatusCode());
+	}
+
+	/**
+	 * @covers /submissions/{submissionId}/participants/{stageId}
+	 * @expectedException GuzzleHttp\Exception\ClientException
+	 */
+// 	public function testGetSubmissionParticipantsAssignedToStageWithInvalidId() {
```
Comment (asmecher) [inline tests/api/SubmissionsTest.php]: (dead code)

File: tests/api/SubmissionsTest.php (hunk 2 above)
Comment (asmecher) [reply tests/api/SubmissionsTest.php]: I'm afraid not -- I haven't worked with any of these tools. But something that validates a response against a Swagger description should be possible -- that link above is just a suggestion.

Comment (NateWr) [reply tests/api/SubmissionsTest.php]: I believe Swagger switched from JSON to YAML a while back and it looks like some of the tooling hasn't caught up. A couple of options come to mind:

1. Parse YAML into PHP and convert to JSON. http://php.net/manual/en/function.yaml-parse.php Will that work?

2. Parse YAML into PHP and roll our own tools to check API responses against valid properties and variable types.

3. Check out one of the other libraries: https://github.com/Maks3w/SwaggerAssertions or https://github.com/byjg/php-swagger-test

It looks like Postman has a nice toolset for creating, testing and documenting APIs. But it's all hosted in "the cloud", and I think we're on the right track with keeping the documentation under version control in the source. Apparently we can export from Postman to a JSON file. If this happened to match the Swagger style, that could work well.

I haven't found a nice, well-maintained technical-spec-to-user-documentation tool like Swagger's. It seems like we are going to have to close a gap in tooling somewhere. It's probably easiest to do that on the testing side, reading from the YAML (or JSON) and automating the checks.

**Comment by kaschioudi** [comment]
@asmecher . all the tests have passed. this is ready for another review. thanks.
Once merged, I will write the tests for OMP.

**Comment by asmecher** [comment]
Thanks, @kaschioudi. A couple of thoughts in looking this over:
- On IDs: I see some of the tests hard-code database IDs (e.g. submission IDs), some for IDs that should exist and have specific characteristics, and some that shouldn't exist. I'd suggest avoiding hard-coding submission IDs that should exist with some kind of characteristic. It may be possible e.g. to use the web interface to find a submission by title (there's a helper function for this) then pick the ID out of the URL, then use that for the API tests. Other ideas welcome too (the API might support custom URL identifiers, I think), but I don't think we should be hard-coding database IDs that are auto-generated. (Using a high value that is assumed not to exist is OK by me, I think.)
- Configuration. The test suite currently uses system environment variables for details like the base URL. Can you see if it's possible to use those in place of the configuration file you're introducing?

**Comment by NateWr** [comment]
A note on the IDs: you should be able to use the API response to collect a working ID. All summary views of objects should include an `_href` property which points to that object's URL in the API. For example:

```
$response = $this->_sendRequest('GET', '/_submissions');
$data = $this->_getResponseData($response);
$href = $data['items'][0]['_href'];
// $href should equal something like `/submissions/1`
$response = $this->_sendRequest('GET', $href);
$data = $this->_getResponseData($response);
// $data should now represent a submission object
```

**Comment by kaschioudi** [comment]
thanks for you feedback. I have written those tests on the assumption that OJS test data is loaded in the database. Ideally repositories or Active records like objects should have been used to seed DB tables and create some submissions in the setup() method before running the tests. It's because I have not seen those mechanisms that I went with the assumption that OJS test data will be loaded before running REST API tests.

Also note that those IDs were chosen in order to test specific cases (galleys and published state)

> It may be possible e.g. to use the web interface to find a submission by title (there's a helper function for this) then pick the ID out of the URL, then use that for the API tests.

Alright. What's the method name? However with that approach, we may not be hardcoding the IDs but there's still some coupling on the submission title (which will be hardcoded in the source) and the dataset in the database.

> A note on the IDs: you should be able to use the API response to collect a working ID. All summary views of objects should include an _href property which points to that object's URL in the API. 

I am not sure if it's a good practice to rely on the API response to test the API. But I could follow that path to test responses content (if that's okay with Alec). Howerver it's still going to be hard to make assumptions about the submissions states since we don't know much about the returned data by the API since we haven't seed the database ourselves.

**Comment by asmecher** [comment]
I agree with @NateWr that the API is a better approach (since that's essentially what we're testing here).

I don't mind if we bind with titles -- they're easier to maintain/debug than hard-coded numeric IDs.

**Comment by asmecher** [comment]
Yes, we don't currently have a good alternative to that.

**Comment by kaschioudi** [comment]
@asmecher : it seems that these failures are due to network right? 

it so, this is ready for another review.

**Comment by asmecher** [comment]
I'm looking into those this morning.

**Comment by asmecher** [comment]
Is `config.json` still required?

**Comment by kaschioudi** [comment]
> Is config.json still required?

No. I have removed it. Configs values are now read from environment variables (https://github.com/kaschioudi/pkp-lib/blob/b285a0643ad721fc056e053414f92b44590cf442/tests/PKPApiTestCase.inc.php#L65-L67)
and it's documented here (https://github.com/kaschioudi/pkp-lib/blob/2875-fix/tools/runAllTests.sh#L28-L29)

**Comment by kaschioudi** [comment]
@NateWr @asmecher : FYI I came across this small bug while trying to search submissions by title (https://github.com/kaschioudi/pkp-lib/commit/c8fb1d2543bdd4fd3aeefc47d8bad332653ef567)

**Comment by NateWr** [comment]
@kaschioudi where in the UI did you run across the search bug? On the backend we should be using the API/QueryListBuilder, but maybe there's somewhere else it's getitng used.

**Comment by NateWr** [comment]
I think that's ok.

In general, I think we'd like to move more of the logic over to QueryListBuilders. But in this case it's alright. I just wanted to know if there was somewhere in the UI we might have to think about refactoring in the future.

**Pull request closed without merging**